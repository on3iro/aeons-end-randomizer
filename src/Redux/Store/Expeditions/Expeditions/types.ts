import { ActionsUnion } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { actions } from './actions'

export enum ActionTypes {
  CREATE_EXPEDITION = 'Expeditions/Expeditions/CREATE_EXPEDITION',
  CREATE_EXPEDITION_SUCCESS = 'Expeditions/Expeditions/CREATE_EXPEDITION_SUCCESS',
  DELETE_EXPEDITION = 'Expeditions/Expeditions/DELETE_EXPEDITION',
  ROLL_BATTLE = 'Expeditions/Expeditions/ROLL_BATTLE',
  ROLL_BATTLE_SUCCESS = 'Expeditions/Expeditions/ROLL_BATTLE_SUCCESS',
  START_BATTLE = 'Expeditions/Expeditions/START_BATTLE',
  WIN_BATTLE = 'Expeditions/Expeditions/WIN_BATTLE',
  WIN_BATTLE_SUCCESS = 'Expeditions/Expeditions/WIN_BATTLE_SUCCESS',
  LOSE_BATTLE = 'Expeditions/Expeditions/LOSE_BATTLE',
  ROLL_LOSS = 'Expeditions/Expeditions/ROLL_LOSS',
  ROLL_LOSS_SUCCESS = 'Expeditions/Expeditions/ROLL_LOSS_SUCCESS',
  ACCEPT_LOSS = 'Expeditions/Expeditions/ACCEPT_LOSS',
  FINISH_BATTLE = 'Expeditions/Expeditions/FINISH_BATTLE',
  FINISH_EXPEDITION = 'Expeditions/Expeditions/FINISH_EXPEDITION',
  SET_TO_DB = 'Expeditions/Expeditions/SET_TO_DB',
  SET_TO_DB_SUCCESS = 'Expeditions/Expeditions/SET_TO_DB_SUCCESS',
  SET_TO_DB_FAILURE = 'Expeditions/Expeditions/SET_TO_DB_FAILURE',
  FETCH_FROM_DB = 'Expeditions/Expeditions/FETCH_FROM_DB',
  FETCH_FROM_DB_SUCCESS = 'Expeditions/Expeditions/FETCH_FROM_DB_SUCCESS',
  FETCH_FROM_DB_FAILURE = 'Expeditions/Expeditions/FETCH_FROM_DB_FAILURE',
  SETTINGS_SNAPSHOT_MIGRATION_SUCCESS = 'Expeditions/Expeditions/SETTINGS_SNAPSHOT_MIGRATION_SUCCESS',
  EXPEDITION_SHARE = 'Expeditions/Expeditions/SHARE_EXPEDITION',
  EXPEDITION_SHARE_SUCCESS = 'Expeditions/Expeditions/SHARE_EXPEDITION_SUCCESS',
  EXPEDITION_SHARE_FAILURE = 'Expeditions/Expeditions/SHARE_EXPEDITION_FAILURE',
  NARRATIVE_RESOLVE = 'Expeditions/Expeditions/RESOLVE_NARRATIVE',
  REWARDS_GENERATE = 'Expeditions/Expeditions/GENERATE_REWARDS',
  REWARDS_GENERATE_SUCCESS = 'Expeditions/Expeditions/GENERATE_REWARDS_SUCCESS',
  REWARDS_ACCEPT = 'Expeditions/Expeditions/ACCEPT_REWARDS',
}

export type Action = ActionsUnion<typeof actions>

export type State = {
  expeditions: types.Expeditions
  expeditionIds: string[]
}

export type BaseConfig = {
  variantId?: string
  name: string
  bigPocketVariant: boolean
  uniqueMageNames?: boolean
  marketId: string
  expeditionConfig?: types.ImportedExpeditionConfig
  seedValue?: string
}

export type WinConfig = {
  battle: types.Battle
  treasureIds: string[]
  gemIds: string[]
  relicIds: string[]
  spellIds: string[]
}

export type RewardType =
  | 'mage'
  | 'gem'
  | 'relic'
  | 'spell'
  | 'treasure1'
  | 'treasure2'
  | 'treasure3'

export type BattleRewardsResult = types.Battle & {
  seed: types.Seed
}

export type LossConfig = {
  battle: types.Battle
  type: RewardType
  mageIds: string[]
  gemIds: string[]
  relicIds: string[]
  spellIds: string[]
  treasure1Ids: string[]
  treasure2Ids: string[]
  treasure3Ids: string[]
}

export type ExpeditionsStateSlice = {
  Expeditions: {
    Expeditions: {
      expeditions: types.Expeditions
    }
  }
}

export type ExpeditionIdsStateSlice = {
  Expeditions: {
    Expeditions: {
      expeditionIds: string[]
    }
  }
}
