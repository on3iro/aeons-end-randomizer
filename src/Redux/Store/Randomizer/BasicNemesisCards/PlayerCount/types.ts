import { ActionsUnion } from '@martin_hotell/rex-tils'

import * as types from 'aer-types/types'

import { actions } from './actions'

export type State = PlayerCount

export enum ActionTypes {
  SET = 'BasicNemesisCards/PlayerCount/SET',
}

export type Action = ActionsUnion<typeof actions>

export type PlayerCount = types.PlayerCount

export type PlayerCountStateSlice = {
  Randomizer: {
    BasicNemesisCards: {
      PlayerCount: PlayerCount
    }
  }
}
