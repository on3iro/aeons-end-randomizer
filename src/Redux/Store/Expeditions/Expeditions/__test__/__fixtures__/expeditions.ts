export const expedition1 = {
  id: 'Expedition1',
  name: 'test',
  score: 4,
  seed: {
    seed: 'Expedition1',
    supplyState: true,
    nemesisState: true,
  },
  settingsSnapshot: {
    supplySetup: {
      id: 'random',
      name: 'Random Setup (Default)',
      type: 'official',
      default: true,
      active: true,
      tiles: [
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Relic',
          operation: 'ANY',
        },
        {
          type: 'Relic',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
      ],
    },
    availableCardIds: [
      'DiamondCluster',
      'ChaosArc',
      'Ignite',
      'EssenceTheft',
      'SearingRuby',
      'FeralLightning',
      'PlanarInsight',
      'SpectralEcho',
      'BurningOpal',
      'ConsumingVoid',
      'UnstablePrism',
      'CloudedSapphire',
      'MagesTalisman',
      'FlexingDagger',
      'LavaTendril',
      'BottledVortex',
      'ArcaneNexus',
      'DarkFire',
      'PhoenixFlame',
      'Jade',
      'AmplifyVision',
      'VriswoodAmber',
      'BlastingStaff',
      'SiftersPearl',
      'WildfireWhip',
      'FocusingOrb',
      'OblivionSwell',
      'TripliteCore',
      'NerveJab',
      'DoubleTap',
      'ObliviumResin',
      'Summonite',
      'ManifoldContainer',
      'DrainingTouch',
      'JoltingCrust',
      'MentiteChunk',
      'SoulCords',
      'VoidiumSpike',
      'PrecisionShot',
      'MutedLacosite',
      'BladeOfWisdom',
      'StunningForce',
      'FlashOfIntellect',
      'ScholarsOpus',
      'BottledStar',
      'BurningCurrent',
      'PsychicEruption',
      'LivingGauntlet',
      'ShiningTetrite',
      'VoidBond',
      'Combustion',
      'VimDynamo',
      'DisintegratingScythe',
      'MonstrousInferno',
      'DevouringShadow',
      'BanishingTopaz',
      'Transmogrifier',
      'JeweledBrain',
      'BlastSphere',
      'WoundMender',
      'LurkingSionite',
      'UnrefinedBlaststone',
      'Exogranite',
      'RecurringJasper',
      'EnergizedRubidium',
      'ResonantPearl',
      'CagedFire',
      'GalvanizedBauble',
      'AetherConduit',
      'WellOfEnergy',
      'PainConduit',
      'LinkConduit',
      'MarbleGalaxy',
      'DelugeOfPower',
      'ReverberatingShock',
      'FatalHarmony',
      'BouncingBoom',
      'SpiritLift',
      'Tethered Darts',
      'ForceAmplifier',
      'ImbuedSmash',
      'PatternStrike',
      'MoltenHammer',
      'Radiance',
      'Blaze',
      'ScryingBolt',
      'SagesBrand',
      'LeechingAgate',
      'TemporalHelix',
      'Char',
      'AlienElement',
      'Scorch',
      'Pyromancy',
      'FeedbackAura',
      'Catalyst',
      'PainStone',
      'AstralCube',
      'RiddleSphere',
      'NetherConduit',
      'HauntedBerylite',
      'Olivinite',
      'CacheGlass',
      'ReflectiveConduit',
      'WillWeaver',
      'BreachSeeker',
      'CrescendoRay',
      'StormVapors',
      'MantraOfStrength',
      'EmbodyFlame',
      'SnapRitual',
      'ShiningFluorite',
      'PolyphaseTurbine',
      'BreachExtractor',
      'FocusingConduit',
      'DualFlash',
      'DustCaller',
      'SymbioticSynapse',
      'FossilizedScarab',
      'ThermalDart',
      'Resonate',
      'Conflagration',
      'DimensionalKey',
      'InnerFire',
      'Fulminate',
      'EternityCharm',
      'ErraticIngot',
      'ReduceToAsh',
      'VortexGauntlet',
      'Carbonize',
      'Equilibrium',
      'VolcanicGlass',
      'ConvectionField',
      'JaggedLightning',
      'BloodstoneJewel',
      'Kindle',
      'ThoughtformFamiliar',
      'Crystallize',
      'FieryTorrent',
      'CelestialSpire',
      'ConjureTheLost',
      'NovaForge',
      'DreadDiamond',
      'CairnCompass',
      'ScoriaSlag',
      'BreachOre',
      'FiendCatcher',
      'PyrotechnicSurge',
      'Aurora',
      'ConclaveScroll',
      'PrimordialFetish',
      'FrozenMagmite',
      'MagesTotem',
      'DrownInFlames',
      'FleetingVision',
      'SplinterMissile',
      'EchoStone',
      'ThievingSpirit',
      'FracturedQuartz',
      'RockLauncher',
      'AbacusOfIgnition',
      'IlluminatingFlame',
      'ChronalArc',
    ],
    availableMageIds: [
      'Adelheim',
      'Brama',
      'Jian',
      'Kadir',
      'Lash',
      'Mist',
      'Phaedraxa',
      'Xaxos',
      'Nym',
      'Reeve',
      'Zhana',
      'Inco',
      'Razra',
      'Sahala',
      'Gygar',
      'Soskel',
      'Taqren',
      'Talix',
      'Rhia',
      'Claudia',
      'Lost',
      'Malastar',
      'Indira',
      'Remnant',
      'Nook',
      'Mazra',
      'Qu',
      'Sparrow',
      'XaxosTV',
      'Dezmodia',
      'Garu',
      'Gex',
      'Mazahaedron',
      'MistWE',
      'Quilius',
      'Ulgimor',
      'YanMagda',
    ],
    availableNemesisIds: [
      'CarapaceQueen',
      'CrookedMask',
      'PrinceOfGluttons',
      'Rageborne',
      'HordeCrone',
      'Necroswarm',
      'Maggoth',
      'Arachnos',
      'AgelessWalker',
      'Fenrix',
      'BlightLord',
      'WaywardOne',
      'ThriceDeadProphet',
      'Wraithmonger',
      'TheWailing',
      'TheWanderer',
      'KnightOfShackles',
      'MaidenOfThorns',
      'GateWitch',
      'HollowCrown',
      'MagusOfCloaks',
      'UmbraTitan',
    ],
    availableTreasureIds: [
      'DezmodiasOblivionShard',
      'UlgimorsCoalFragment',
      'IndirasLoneOpal',
      'MistsAmethystParagon',
      'MalastarsImmolate',
      'YanMagdasIlluminate',
      'EyeOfTheMaelstrom',
      'WellOfDespair',
      'CronesAmulet',
      'EdibleFungusChunks',
      'UmbralHornOfWar',
      'TheBrokenMask',
      'AlchemistsAlembic',
      'SoothsayersPouch',
      'SuppressingPin',
      'Quicksilver',
      'BandOfRetrieval',
      'PurifiedBangle',
      'KadirsEmeraldShard',
      'GexsShatteredGeode',
      'JiansMoonstoneShard',
      'LashsQuartzShard',
      'XaxosFlare',
      'AdelheimsAmethystShard',
      'BramasBuriedLight',
      'GarusTorch',
      'XaxosPyre',
      'MistsGarnetShard',
      'ShimmeringCloakOfTheMagus',
      'ImbuedShackles',
      'GluttonsTooth',
      'CoreOfRage',
      'CleansingAmulet',
      'CarapaceFragement',
      'ThornedWhip',
      'BonesOfDeathmind',
      'AcceleratingGauntlets',
      'EssenceExtractor',
      'BroochOfAttunment',
      'BladedCrystal',
      'PrismOfDestruction',
      'ForgottenTrinket',
      'PrecisionMagnifier',
      'ShroudOfObfucation',
      'VerdantStaff',
      'FlowingMantle',
      'MazahaedronsWorldheartShard',
      'PhaedraxasTourmalineShard',
      'FracturedShell',
      'Extra-DimensionalLens',
      'SeersBracer',
      'RemnantsVoidShard',
      'QuiliussErasure',
      'ZhanasFleetingEmber',
      'BlightedRootClump',
      'PetrifiedWitchFinger',
      'WaywardScraps',
      'TrueSightMonocle',
      'EntwinedAetherStrands',
      'SiphoningBlade',
      'GemEncrustedAnklet',
      'ReevesRavenShard',
      'NymsVision',
      'SparrowsAid',
      'WraithsEssence',
      'VolatileClasp',
      'BreachStabilizer',
      'FangedChoker',
      'EndlessBandolier',
    ],
    availableUpgradedBasicNemesisCardIds: [
      'HissingAcid',
      'Wreck',
      'BaneCommander',
      'SkyTremor',
      'EngulfingMadness',
      'Separate',
      'NeedleDoom',
      'Marked',
      'Eradicate',
      'FlashOfDecay',
      'Terrify',
      'RayOfDesolation',
      'ThunderousTempest',
      'Sacrifice',
      'Burialskulk',
      'Rupture',
      'FurrowOfDestruction',
      'Mangle',
      'HowlingSlayer',
    ],
    usedExpansions: [
      'AE',
      'BS',
      'Depths',
      'IW',
      'NA',
      'Nameless',
      'OD',
      'SD',
      'TA',
      'TV',
      'WE',
      'promos',
    ],
  },
  barracks: {
    mageIds: ['Qu', 'Mazra', 'Nook', 'Sparrow', 'Gygar'],
    supplyIds: [
      'DiamondCluster',
      'PainConduit',
      'UnstablePrism',
      'TemporalHelix',
      'ThoughtformFamiliar',
      'BurningCurrent',
      'StormVapors',
      'CrescendoRay',
      'Aurora',
    ],
    treasureIds: [
      'SparrowsAid',
      'MistsAmethystParagon',
      'BramasBuriedLight',
      'XaxosPyre',
      'IndirasLoneOpal',
    ],
  },
  upgradedBasicNemesisCards: [],
  banished: ['SearingRuby', 'RecurringJasper', 'PainStone'],
  variantId: 'DEFAULT',
  bigPocketVariant: false,
  sequence: {
    firstBattleId: 'CvghslX3mK',
    branches: {
      CvghslX3mK: {
        id: 'CvghslX3mK',
        nextBranchId: 'YBsJLx4PfV',
        type: 'battle',
        expeditionId: 'Expedition1',
        nemesisId: 'KnightOfShackles',
        config: {
          tier: 1,
          newUBNCards: {
            ids: [],
            addRandom: false,
          },
          treasure: {
            level: 1,
            hasTreasure: true,
          },
        },
        status: 'finished',
        tries: 2,
        rewards: {
          treasure: [
            'SparrowsAid',
            'MistsAmethystParagon',
            'BramasBuriedLight',
            'XaxosPyre',
            'IndirasLoneOpal',
          ],
          supplyIds: ['DiamondCluster', 'PainConduit', 'ThoughtformFamiliar'],
        },
      },
      YBsJLx4PfV: {
        id: 'YBsJLx4PfV',
        nextBranchId: '0t22iBGUff',
        type: 'battle',
        expeditionId: 'Expedition1',
        config: {
          tier: 2,
          newUBNCards: {
            ids: [],
            addRandom: true,
          },
          treasure: {
            level: 2,
            hasTreasure: true,
          },
        },
        status: 'unlocked',
        tries: 0,
      },
      '0t22iBGUff': {
        id: '0t22iBGUff',
        nextBranchId: 'wXBtzLO1bM',
        type: 'battle',
        expeditionId: 'Expedition1',
        config: {
          tier: 3,
          newUBNCards: {
            ids: [],
            addRandom: true,
          },
          treasure: {
            level: 3,
            hasTreasure: true,
          },
        },
        status: 'locked',
        tries: 0,
      },
      wXBtzLO1bM: {
        id: 'wXBtzLO1bM',
        type: 'battle',
        config: {
          tier: 4,
          newUBNCards: {
            ids: [],
            addRandom: true,
          },
          treasure: {
            hasTreasure: false,
          },
        },
        nemesisTier: {
          isNewTier: true,
        },
        expeditionId: 'Expedition1',
        status: 'locked',
        tries: 0,
      },
    },
  },
  battles: [],
  finished: false,
}

export const expedition2 = {
  id: 'Expedition2',
  name: 'test',
  score: 4,
  seed: {
    seed: 'Expedition2',
    supplyState: true,
    nemesisState: true,
  },
  barracks: {
    mageIds: ['Qu', 'Mazra', 'Nook', 'Sparrow', 'Gygar'],
    supplyIds: [
      'DiamondCluster',
      'PainConduit',
      'UnstablePrism',
      'TemporalHelix',
      'ThoughtformFamiliar',
      'BurningCurrent',
      'StormVapors',
      'CrescendoRay',
      'Aurora',
    ],
    treasureIds: [
      'SparrowsAid',
      'MistsAmethystParagon',
      'BramasBuriedLight',
      'XaxosPyre',
      'IndirasLoneOpal',
    ],
  },
  upgradedBasicNemesisCards: [],
  banished: ['SearingRuby', 'RecurringJasper', 'PainStone'],
  variantId: 'DEFAULT',
  bigPocketVariant: false,
  battles: [
    {
      id: 'CvghslX3mK',
      nemesisTier: {
        tier: 1,
        isNewTier: false,
      },
      treasure: {
        level: 1,
        hasTreasure: true,
      },
      expeditionId: 'Expedition2',
      status: 'finished',
      tries: 2,
      nemesisId: 'KnightOfShackles',
      rewards: {
        treasure: [
          'SparrowsAid',
          'MistsAmethystParagon',
          'BramasBuriedLight',
          'XaxosPyre',
          'IndirasLoneOpal',
        ],
        supplyIds: ['DiamondCluster', 'PainConduit', 'ThoughtformFamiliar'],
      },
    },
    {
      id: 'YBsJLx4PfV',
      nemesisTier: {
        tier: 2,
        isNewTier: true,
      },
      treasure: {
        level: 2,
        hasTreasure: true,
      },
      expeditionId: 'Expedition2',
      status: 'unlocked',
      tries: 0,
    },
    {
      id: '0t22iBGUff',
      nemesisTier: {
        tier: 3,
        isNewTier: true,
      },
      treasure: {
        level: 3,
        hasTreasure: true,
      },
      expeditionId: 'Expedition2',
      status: 'locked',
      tries: 0,
    },
    {
      id: 'wXBtzLO1bM',
      nemesisTier: {
        tier: 4,
        isNewTier: true,
      },
      treasure: {
        hasTreasure: false,
      },
      expeditionId: 'Expedition2',
      status: 'locked',
      tries: 0,
    },
  ],
  finished: false,
}
