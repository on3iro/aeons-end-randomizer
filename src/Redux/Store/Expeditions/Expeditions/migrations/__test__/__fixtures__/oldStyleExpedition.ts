import * as types from 'types'

export const oldStyleExpedition: types.OldStyleExpedition = {
  id: 'OLD_STYLE',
  name: 'Oldstyle',
  score: 6,
  seed: {
    seed: 'OLD_STYLE',
    supplyState: true,
    nemesisState: true,
  },
  settingsSnapshot: {
    supplySetup: {
      id: 'random',
      name: 'Random Setup (Default)',
      type: 'official',
      default: true,
      active: true,
      tiles: [
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Gem',
          operation: 'ANY',
        },
        {
          type: 'Relic',
          operation: 'ANY',
        },
        {
          type: 'Relic',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
        {
          type: 'Spell',
          operation: 'ANY',
        },
      ],
    },
    availableCardIds: ['DiamondCluster', 'ChaosArc', 'Ignite'],
    availableMageIds: ['Adelheim', 'Brama', 'Jian'],
    availableNemesisIds: ['CarapaceQueen', 'CrookedMask', 'PrinceOfGluttons'],
    availableTreasureIds: [
      'DezmodiasOblivionShard',
      'UlgimorsCoalFragment',
      'IndirasLoneOpal',
    ],
    availableUpgradedBasicNemesisCardIds: [
      'HissingAcid',
      'Wreck',
      'BaneCommander',
    ],
    usedExpansions: ['AE'],
  },
  barracks: {
    mageIds: ['Reeve', 'Zhana', 'Inco', 'Nym'],
    supplyIds: [
      'ArcingSilicate',
      'SearingRuby',
      'ObliviumResin',
      'BurningOpal',
      'RockLauncher',
      'WarpingHaze',
      'Fulminate',
      'PyroGeist',
      'Char',
    ],
    treasureIds: [
      'GexsShatteredGeode',
      'ReevesRavenShard',
      'XaxosPyre',
      'SparrowsAid',
      'XaxosFlare',
    ],
  },
  upgradedBasicNemesisCards: [],
  banished: ['VimDynamo', 'FocusingOrb', 'Tethered Darts'],
  variantId: 'DEFAULT',
  bigPocketVariant: false,
  battles: [
    {
      id: 'fHFX3VH99',
      nemesisTier: {
        tier: 1,
        isNewTier: false,
      },
      treasure: {
        level: 1,
        hasTreasure: true,
      },
      expeditionId: 'OLD_STYLE',
      status: 'finished',
      tries: 1,
      nemesisId: 'Bladius',
      rewards: {
        treasure: [
          'GexsShatteredGeode',
          'ReevesRavenShard',
          'XaxosPyre',
          'SparrowsAid',
          'XaxosFlare',
        ],
        supplyIds: ['ObliviumResin', 'RockLauncher', 'Char'],
      },
    },
    {
      id: 'dmvgVnT3c',
      nemesisTier: {
        tier: 2,
        isNewTier: true,
      },
      treasure: {
        level: 2,
        hasTreasure: true,
      },
      expeditionId: 'OLD_STYLE',
      status: 'unlocked',
      tries: 0,
    },
    {
      id: 'olps4itk4',
      nemesisTier: {
        tier: 3,
        isNewTier: true,
      },
      treasure: {
        level: 3,
        hasTreasure: true,
      },
      expeditionId: 'OLD_STYLE',
      status: 'locked',
      tries: 0,
    },
    {
      id: 'y0gQDPjY5',
      nemesisTier: {
        tier: 4,
        isNewTier: true,
      },
      treasure: {
        hasTreasure: false,
      },
      expeditionId: 'OLD_STYLE',
      status: 'locked',
      tries: 0,
    },
  ],
  finished: false,
  migrationVersion: 2020030301,
}
