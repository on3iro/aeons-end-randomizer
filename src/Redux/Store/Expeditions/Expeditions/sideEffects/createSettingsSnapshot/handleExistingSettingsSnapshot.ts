import * as types from 'aer-types/types'
import { determineUsedExpansions } from './determineUsedExpansions'

import { MagesContentStateSlice } from 'Redux/Store/Settings/Expansions/Mages/content'
import { NemesisContentStateSlice } from 'Redux/Store/Settings/Expansions/Nemeses/content'
import { TreasureContentStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/content'
import { UpgradedBasicNemesisCardContentStateSlice } from 'Redux/Store/Settings/Expansions/UpgradedBasicNemesisCards/content'
import { CardsContentStateSlice } from 'Redux/Store/Settings/Expansions/Cards/content'

export const handleExistingSettingsSnapshot = (
  state: CardsContentStateSlice &
    MagesContentStateSlice &
    NemesisContentStateSlice &
    TreasureContentStateSlice &
    UpgradedBasicNemesisCardContentStateSlice,
  existingConfig: types.ExpeditionConfig,
  baseSupplySetups: { [id: string]: types.IMarketSetup },
  marketId: string
): types.SettingsSnapshot => {
  const usedExpansions = determineUsedExpansions(
    state,
    existingConfig.settingsSnapshotConfig,
    existingConfig.sequenceConfig.branches,
    existingConfig.initialBarracksConfig
  )

  // NOTE: Because we currently do get the marketId from the outside
  // (it is currently passed to an action creator from a react component),
  // we do not know the actual supply setup.
  // Therefore we create a lookup object from all predefined setups, user specific custom setups
  // as well as the setup which was used inside the given snapshot and retrieve the setup by its id.
  //
  // FIXME: This should probably be refactored in the future.
  const existingSupply = {
    [existingConfig.settingsSnapshotConfig.supplySetup.id]:
      existingConfig.settingsSnapshotConfig.supplySetup,
  }

  const allSupplySetups = {
    ...baseSupplySetups,
    ...existingSupply,
  }

  const supplySetup = allSupplySetups[marketId]

  return {
    ...existingConfig.settingsSnapshotConfig,
    usedExpansions,
    supplySetup,
  }
}
