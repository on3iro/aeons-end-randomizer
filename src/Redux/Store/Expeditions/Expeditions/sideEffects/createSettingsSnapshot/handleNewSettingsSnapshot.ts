import * as types from 'aer-types/types'
import { determineUsedExpansions } from './determineUsedExpansions'

import { MagesContentStateSlice } from 'Redux/Store/Settings/Expansions/Mages/content'
import { NemesisContentStateSlice } from 'Redux/Store/Settings/Expansions/Nemeses/content'
import { TreasureContentStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/content'
import { UpgradedBasicNemesisCardContentStateSlice } from 'Redux/Store/Settings/Expansions/UpgradedBasicNemesisCards/content'
import { CardsContentStateSlice } from 'Redux/Store/Settings/Expansions/Cards/content'

export const handleNewSettingsSnapshot = (
  state: CardsContentStateSlice &
    MagesContentStateSlice &
    NemesisContentStateSlice &
    TreasureContentStateSlice &
    UpgradedBasicNemesisCardContentStateSlice,
  baseSnapshot: {
    supplySetup: types.IMarketSetup
    availableCardIds: string[]
    availableMageIds: string[]
    availableNemesisIds: string[]
    availableTreasureIds: string[]
    availableUpgradedBasicNemesisCardIds: string[]
  }
): types.SettingsSnapshot => {
  const usedExpansions = determineUsedExpansions(state, baseSnapshot)

  return {
    ...baseSnapshot,
    usedExpansions,
  }
}
