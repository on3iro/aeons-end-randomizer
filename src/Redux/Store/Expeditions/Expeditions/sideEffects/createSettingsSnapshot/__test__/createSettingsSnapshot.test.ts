import { createSettingsSnapshot } from '../index'
import * as handleExistingSettingsSnapshotModule from '../handleExistingSettingsSnapshot'
import * as handleNewSettingsSnapshotModule from '../handleNewSettingsSnapshot'
import * as types from 'aer-types/types'

import { rootState } from '__fixtures__/rootState'

describe('createSettingsSnapshot()', () => {
  it('should call handleNewSettingsSnapshot()', () => {
    const handleNewSettingsSnapshotSpy = jest.spyOn(
      handleNewSettingsSnapshotModule,
      'handleNewSettingsSnapshot'
    )

    const handleExistingSettingsSnapshotSpy = jest.spyOn(
      handleExistingSettingsSnapshotModule,
      'handleExistingSettingsSnapshot'
    )

    createSettingsSnapshot(rootState, undefined)

    expect(handleNewSettingsSnapshotSpy).toHaveBeenCalled()
    expect(handleExistingSettingsSnapshotSpy).not.toHaveBeenCalled()

    handleNewSettingsSnapshotSpy.mockRestore()
    handleExistingSettingsSnapshotSpy.mockRestore()
  })

  it('should call handleExistingSettingsSnapshot()', () => {
    const handleNewSettingsSnapshotSpy = jest.spyOn(
      handleNewSettingsSnapshotModule,
      'handleNewSettingsSnapshot'
    )

    const handleExistingSettingsSnapshotSpy = jest.spyOn(
      handleExistingSettingsSnapshotModule,
      'handleExistingSettingsSnapshot'
    )

    createSettingsSnapshot(rootState, {
      name: 'TEST',
      bigPocketVariantConfig: true,
      sequenceConfig: {
        firstBranchId: '',
        branches: {} as types.BranchesConfig,
      },
      settingsSnapshotConfig: {
        supplySetup: {
          id: 'random',
          name: 'Random Setup (Default)',
          type: 'official',
          default: true,
          active: true,
          tiles: [
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Relic',
              operation: 'ANY',
            },
            {
              type: 'Relic',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
          ],
        },
        availableCardIds: [
          'DiamondCluster',
          'ChaosArc',
          'Ignite',
          'EssenceTheft',
          'SearingRuby',
        ],
        availableMageIds: ['Adelheim'],
        availableNemesisIds: ['CarapaceQueen', 'CrookedMask'],
        availableTreasureIds: [
          'DezmodiasOblivionShard',
          'UlgimorsCoalFragment',
        ],
        availableUpgradedBasicNemesisCardIds: [
          'HissingAcid',
          'Wreck',
          'BaneCommander',
        ],
      },
    })

    expect(handleNewSettingsSnapshotSpy).not.toHaveBeenCalled()
    expect(handleExistingSettingsSnapshotSpy).toHaveBeenCalled()

    handleNewSettingsSnapshotSpy.mockRestore()
    handleExistingSettingsSnapshotSpy.mockRestore()
  })

  it('should produce correct result', () => {
    const result = createSettingsSnapshot(rootState, {
      name: 'TEST',
      bigPocketVariantConfig: true,
      sequenceConfig: {
        firstBranchId: '',
        branches: {} as types.BranchesConfig,
      },
      settingsSnapshotConfig: {
        supplySetup: {
          id: 'random',
          name: 'Random Setup (Default)',
          type: 'official',
          default: true,
          active: true,
          tiles: [
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Gem',
              operation: 'ANY',
            },
            {
              type: 'Relic',
              operation: 'ANY',
            },
            {
              type: 'Relic',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
            {
              type: 'Spell',
              operation: 'ANY',
            },
          ],
        },
        availableCardIds: [
          'DiamondCluster',
          'ChaosArc',
          'Ignite',
          'EssenceTheft',
          'SearingRuby',
        ],
        availableMageIds: ['Adelheim'],
        availableNemesisIds: ['CarapaceQueen', 'CrookedMask'],
        availableTreasureIds: [
          'DezmodiasOblivionShard',
          'UlgimorsCoalFragment',
        ],
        availableUpgradedBasicNemesisCardIds: [
          'HissingAcid',
          'Wreck',
          'BaneCommander',
        ],
      },
    })

    expect(result).toMatchSnapshot()
  })
})
