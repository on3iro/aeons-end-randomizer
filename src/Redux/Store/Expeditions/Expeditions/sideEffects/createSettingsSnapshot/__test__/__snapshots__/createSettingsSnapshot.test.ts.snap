// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSettingsSnapshot() should produce correct result 1`] = `
Object {
  "availableCardIds": Array [
    "DiamondCluster",
    "ChaosArc",
    "Ignite",
    "EssenceTheft",
    "SearingRuby",
  ],
  "availableMageIds": Array [
    "Adelheim",
  ],
  "availableNemesisIds": Array [
    "CarapaceQueen",
    "CrookedMask",
  ],
  "availableTreasureIds": Array [
    "DezmodiasOblivionShard",
    "UlgimorsCoalFragment",
  ],
  "availableUpgradedBasicNemesisCardIds": Array [
    "HissingAcid",
    "Wreck",
    "BaneCommander",
  ],
  "supplySetup": Object {
    "active": true,
    "default": true,
    "id": "random",
    "name": "Random Setup (Default)",
    "tiles": Array [
      Object {
        "operation": "ANY",
        "type": "Gem",
      },
      Object {
        "operation": "ANY",
        "type": "Gem",
      },
      Object {
        "operation": "ANY",
        "type": "Gem",
      },
      Object {
        "operation": "ANY",
        "type": "Relic",
      },
      Object {
        "operation": "ANY",
        "type": "Relic",
      },
      Object {
        "operation": "ANY",
        "type": "Spell",
      },
      Object {
        "operation": "ANY",
        "type": "Spell",
      },
      Object {
        "operation": "ANY",
        "type": "Spell",
      },
      Object {
        "operation": "ANY",
        "type": "Spell",
      },
    ],
    "type": "official",
  },
  "usedExpansions": Array [
    "AE",
    "IW",
    "NA",
  ],
}
`;
